File: pkg/R/rank1_update.R
## ID: rank1_update.R, last updated 2026-01-12, F.Osorio

rank1.update <- function(a, alpha, u, diag = FALSE)
{ ## Rank-1 undate: mat <- a + alpha * u %*% t(u)
  ## if diag = TRUE, mat <- diag(a) + alpha * u %*% t(u)
  if (is.data.frame(a))
    a <- as.matrix(a)
  if (!is.numeric(a))
    stop("'a' is not a numeric argument")
  if (diag) {
    job <- 0
    if (is.matrix(a))
      a <- diag(a)
    if (!is.vector(a))
      stop("supply a vector-like 'a'")
    n <- length(a)
  } else {
    job <- 1
    if (!is.matrix(a))
      stop("supply a matrix-like 'a'")
    da <- dim(a)
    n <- da[1]
    p <- da[2]
    if (n != p)
      stop("argument 'a' is not a square matrix")
  }

  storage.mode(a) <- "double"
  z <- .C("rank1_update",
          mat = double(n * n),
          ldmat = as.integer(n),
          n = as.integer(n),
          job = as.integer(job),
          a = a,
          alpha = as.double(alpha),
          u = as.double(u))$mat
  z <- matrix(z, nrow = n, ncol = n)
  z
}

File: pkg/src/rank1_update.c
/* $ID: rank1_update.c, last updated 2026-01-13, F.Osorio */

#include "fastmatrix.h"

void
rank1_update(double *mat, int *ldmat, int *n, int *job, double *a, double *alpha, double *u)
{ /* Rank-1 update (to be called by R) */
  int p = *n, task = *job;

  if (task) {
    FM_copy_mat(mat, *ldmat, a, p, p, p);
    if (*alpha == 0.0)
      return;
    BLAS2_ger(*alpha, mat, *ldmat, p, p, u, 1, u, 1);
  }
  else 
    FM_diag_plus_rank1(mat, *ldmat, p, a, *alpha, u);
}

void 
FM_diag_plus_rank1(double *mat, int ldmat, int n, double *a, double alpha, double *u)
{ /* rank 1 update: mat <- diag(a) + alpha * u %*% t(u) (symbol exported to API) */

  if (n <= 0) /*quick return */
    return;

  if (alpha == 0.0) {
    for (int i = 0; i < n; i++)
      mat[i * (ldmat + 1)] = a[i];
    return;
  }

  for (int i = 0; i < n; i++) {
    mat[i * (ldmat + 1)] = a[i] + alpha * SQR(u[i]);
    for (int j = i + 1; j < n; j++) {
      *(mat + i + j * n) = alpha * u[i] * u[j];
      *(mat + j + i * n) = *(mat + i + j * n);
    }
  }
}
